sklearn/_build_utils/__init__.py

```diff
@@ -33,3 +35,42 @@ def atlas_not_found(blas_info_):
         cblas_libs = blas_info.pop('libraries', [])
 
     return cblas_libs, blas_info
+
+
+def get_cython_source(filename):
+    is_c_filename = filename.endswith('.c')
+    is_cpp_filename = filename.endswith('.cpp')
+
+    # files in src are .c and .cpp files that are not cython-generated
+    if ('src' + os.sep) in filename and (is_c_filename or is_cpp_filename):
```

Comment:
This is the most brittle part of the PR I think. I don't know whether there is a better way of doing it, suggestions welcome.

I guess one possible improvement would be to add an additional argument in add_cython_extension like cython_generated_sources so that we can discriminate more easily betwen the .c files coming from cython and the ones in the repo.